corehq/apps/userreports/tasks.py

```diff
@@ -16,65 +16,90 @@
-            rev = 'static'
-        else:
-            rev = config._rev
-        return 'ucr_queue-{}:{}'.format(config._id, rev)
-
-    is_static = _is_static(indicator_config_id)
-    if is_static:
-        config = StaticDataSourceConfiguration.by_id(indicator_config_id)
+def _is_static(config_id):
+    return config_id.startswith(StaticDataSourceConfiguration._datasource_id_prefix)
+
+
+def _get_config_by_id(indicator_config_id):
+    if _is_static(indicator_config_id):
+        return StaticDataSourceConfiguration.by_id(indicator_config_id)
     else:
-        config = DataSourceConfiguration.get(indicator_config_id)
+        return DataSourceConfiguration.get(indicator_config_id)
 
-    adapter = IndicatorSqlAdapter(config)
 
+def _get_redis_key_for_config(config):
+    if _is_static(config._id):
+        rev = 'static'
+    else:
+        rev = config._rev
+    return 'ucr_queue-{}:{}'.format(config._id, rev)
+
+
+def _build_indicators(indicator_config_id, relevant_ids):
+    config = _get_config_by_id(indicator_config_id)
+    adapter = IndicatorSqlAdapter(config)
     couchdb = _get_db(config.referenced_doc_type)
-    client = get_redis_client().client.get_client()
+    redis_client = get_redis_client().client.get_client()
     redis_key = _get_redis_key_for_config(config)
 
-    if len(client.smembers(redis_key)) > 0:
-        relevant_ids = client.smembers(redis_key)
-    else:
-        if not is_static:
-            # Save the start time now in case anything goes wrong. This way we'll be
-            # able to see if the rebuild started a long time ago without finishing.
-            config.meta.build.initiated = datetime.datetime.utcnow()
-            config.save()
-            redis_key = _get_redis_key_for_config(config)
-
-        adapter.rebuild_table()
-        relevant_ids = get_doc_ids_in_domain_by_type(config.domain, config.referenced_doc_type,
-                                   database=couchdb)
-        for docs in chunked(relevant_ids, 1000):
-            client.sadd(redis_key, *docs)
-
     for doc in iter_docs(couchdb, relevant_ids, chunksize=500):
         try:
             # save is a noop if the filter doesn't match
             adapter.save(doc)
-            client.srem(redis_key, doc.get('_id'))
+            redis_client.srem(redis_key, doc.get('_id'))
         except DataError as e:
             logging.exception('problem saving document {} to table. {}'.format(doc['_id'], e))
 
-    if not is_static:
-        client.delete(redis_key)
-        config.meta.build.finished = True
-        try:
-            config.save()
-        except ResourceConflict:
-            current_config = DataSourceConfiguration.get(config._id)
-            # check that a new build has not yet started
-            if config.meta.build.initiated == current_config.meta.build.initiated:
-                current_config.meta.build.finished = True
-                current_config.save()
+        if not _is_static(indicator_config_id):
```

Comment:
I think this if statement needs to be pulled out of this function. We shouldn't mark build as finished for every single doc :)